### PR TITLE
lxc-to-lxd: switch to using whitelist

### PR DESCRIPTION
**Pull request by brauner**
We keep a whitelist for supported configuration keys. We then perform a check
whether the container sets any unsupported configuration keys. We report the
first unsupported configuration key we found back to the user and the error out.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by brauner** [comment]
Closes  #2676.
@tych0, @stgraber, please check for any config keys I might have disallowed that should be allowed and vica versa. :)

**Comment by stgraber** [comment]
I suspect the list should be much more restricted since we don't actually write to raw.lxc for any of the options that don't have a LXD equivalent since we don't really support raw.lxc in LXD.

The exception to that rule being lxc.aa_profile which we do write through raw.lxc.

For the rest, you don't seem to be properly supporting the "lxc.network.0.name" syntax.

**Comment by brauner** [comment]
> I suspect the list should be much more restricted since we don't actually write to raw.lxc for any of the options that don't have a LXD equivalent since we don't really support raw.lxc in LXD.

> The exception to that rule being lxc.aa_profile which we do write through raw.lxc.

Right.

> For the rest, you don't seem to be properly supporting the "lxc.network.0.name" syntax.

Hm? The check `d.startswith(`lxc.network`) should let those through, right (If that's what you're getting at.).

File: scripts/lxc-to-lxd [hunk 1]
```diff
@@ -9,6 +9,80 @@ import subprocess
...
+        # 'lxc.id_map',
+        'lxc.loglevel',
+        # 'lxc.logfile',
+        'lxc.mount.entry',
+        'lxc.mount.auto',
+        'lxc.mount',
+        # 'lxc.rootfs.mount',
+        # 'lxc.rootfs.options',
+        # 'lxc.rootfs.backend',
+        'lxc.rootfs',
```
Comment (tych0) [inline scripts/lxc-to-lxd]: Does it make sense to have this one here?

Comment (brauner) [inline scripts/lxc-to-lxd]: Well it is checked further down in the code (me thinks). :)

Comment (brauner) [inline scripts/lxc-to-lxd]: Also, every container will have this set. So if we kick it, we will not import any container at all. :)

File: scripts/lxc-to-lxd [hunk 2]
```diff
@@ -9,6 +9,80 @@ import subprocess
...
+        # 'lxc.console',
+        # 'lxc.seccomp',
+        'lxc.include',
+        'lxc.autodev',
+        'lxc.haltsignal',
+        'lxc.rebootsignal',
+        'lxc.stopsignal',
+        'lxc.start.auto',
+        'lxc.start.delay',
+        'lxc.start.order',
```
Comment (brauner) [inline scripts/lxc-to-lxd]: Are you sure? The script further down makes use of the values set for those keys. It initializes:

```
boot.autostart
boot.autostart.delay
boot.autostart.priority
```

Comment (stgraber) [inline scripts/lxc-to-lxd]: Oh, forgot I did that, nevermind :)

File: scripts/lxc-to-lxd (hunk 1 above)
Comment (stgraber) [inline scripts/lxc-to-lxd]: Makes sense

File: scripts/lxc-to-lxd [hunk 3]
```diff
@@ -128,6 +212,16 @@ def convert_container(lxd_socket, container_name, args):
...
     print("Parsing LXC configuration")
     lxc_config = config_parse(container.config_file_name)
+    found_keys = config_keys(lxc_config)
+
+    print("Checking for unsupported LXC configuration keys")
+    diff = list(set(found_keys) - set(supported_keys))
+    for d in diff:
+        if (not d.startswith('lxc.network.') and not
+                d.startswith('lxc.cgroup.')):
+            print("Found at least one unsupported config key: ", d)
```
Comment (stgraber) [inline scripts/lxc-to-lxd]: Nitpick, but I think in the rest of the file we use the "%s" % d syntax.

Comment (stgraber) [inline scripts/lxc-to-lxd]: Also, the skip messages are usually one line, so I think we'd want:

    print("Found at least one unsupported config key, "%s", skipping..." % d)

File: scripts/lxc-to-lxd [hunk 4]
```diff
@@ -152,18 +246,6 @@ def convert_container(lxd_socket, container_name, args):
         print("Container name doesn't match lxc.utsname, skipping...")
         return False
 
-    # Detect privileged containers
-    print("Validating container mode")
-    if config_get(lxc_config, "lxc.id_map"):
-        print("Unprivileged containers aren't supported, skipping...")
-        return False
```
Comment (stgraber) [inline scripts/lxc-to-lxd]: I think we should keep that check and run it before the generic one as the custom error messages makes it worth it (people are unlikely to know what lxc.id_map is).

Comment (brauner) [inline scripts/lxc-to-lxd]: Right, good point.

File: scripts/lxc-to-lxd [hunk 5]
```diff
@@ -35,6 +109,16 @@ def config_get(config, key, default=None):
         return result
 
 
+def config_keys(config):
+    keys = []
+    for line in config:
+        fields = line.split("=", 1)
+        if fields[0].strip():
+            keys.append(fields[0].strip())
```
Comment (stgraber) [inline scripts/lxc-to-lxd]: Ah yeah, we do for comments but don't for non-lxc keys.

**Comment by brauner** [comment]
updated

File: scripts/lxc-to-lxd [hunk 6]
```diff
@@ -457,7 +532,7 @@ parser.add_argument("--all", action="store_true", default=False,
                     help="Import all containers")
 parser.add_argument("--delete", action="store_true", default=False,
                     help="Delete the source container")
-parser.add_argument("--copy-rootfs", action="store_true", default=False,
+parser.add_argument("--copy-rootfs", action="store_true", default=True,
```
Comment (brauner) [inline scripts/lxc-to-lxd]: That's the second commit I added. From the commit message:

"This was hitting me by surprise when using this script. I'd argue that users
much rather expect a copy than a move. It also is more safe: If something goes
wrong users will still have their old container in place to recover."

Comment (brauner) [inline scripts/lxc-to-lxd]: Hm, can we offer copy by default and move as an explicit flag?

**Comment by stgraber** [comment]
The following keys are still allowed but aren't mapped to anything in LXD and so should probably trigger an error:

 - lxc.cgroup (could attempt to convert to limits.* keys at a later time)
 - lxc.loglevel
 - lxc.cap.keep
 - lxc.no_new_privs

The following also aren't converted but may be part of default config so can't cause errors. We may be able to cause an import error if they differ to what LXD uses though:
 - lxc.pts
 - lxc.aa_allow_incomplete
 - lxc.cap.drop
 - lxc.autodev
 - lxc.haltsignal
 - lxc.rebootsignal
 - lxc.stopsignal

**Comment by brauner** [comment]
> The following keys are still allowed but aren't mapped to anything in LXD and so should probably trigger an error:

>  - lxc.cgroup (could attempt to convert to limits.* keys at a later time)

We can disable this key and check for `startswith("lxc.cgroup.")`

> - lxc.loglevel

Disabling.

> - lxc.cap.keep

Disabling.

> - lxc.no_new_privs

Disabling.

> The following also aren't converted but may be part of default config so can't cause errors. We may be able to cause an import error if they differ to what LXD uses though:
> - lxc.pts
> - lxc.aa_allow_incomplete
> - lxc.cap.drop
> - lxc.autodev
> - lxc.haltsignal
> - lxc.rebootsignal
> - lxc.stopsignal

Hm, sure.

**Comment by brauner** [comment]
Totally missed mentioning that I updated a few days ago.

File: scripts/lxc-to-lxd [hunk 7]
```diff
@@ -9,6 +9,86 @@ import subprocess
...
+    'lxc.environment',
+    # 'lxc.init_cmd',
+    # 'lxc.init_uid',
+    # 'lxc.init_gid',
+    # 'lxc.ephemeral',
+    # 'lxc.syslog',
+    # 'lxc.no_new_privs',
+
+    # Additional keys that are either set by this script or are used to report
+    # helpful errors to users.
```
Comment (stgraber) [inline scripts/lxc-to-lxd]: The list below should be updated to reflect recent changes.

**Pull request merged**